### DISCUSSION TITLE
Retrieving Stimulus from another trial

### DISCUSSION THREAD

**Discussion by rjenki03** [jspsych/jsPsych#725] [Q&A]
Hello!
I am presenting study word pairs, then asking for a JOL, then a choice of whether to restudy the word pair LATER in the program, pretest the word pair LATER in the program, or neither. 

       var study = {
		type: 'html-keyboard-response',
		stimulus: jsPsych.timelineVariable('stimulus'),		
		};
	var jol = {
		type: 'html-keyboard-response',
		stimulus: ['If we give you the top word of this pair on a test in about 15 minutes,<br>'+
		'what are the chances that you will recall the bottom word?',
		],		
		prompt: 'Type a number between 0 and 10.',
		};
	var choice = {
		type: 'html-keyboard-response',
		stimulus: ['Enter A to study again<br>' +
		           'Enter B to take a practice test<br>' +
				   'Enter C if you are Done',
		],
		choices: [65, 66, 67], //numeric codes for a,b,c		
		};

This is what my stimulus looks like:

	var test_stimuli = [
		{ stimulus: 'bowl' + '<br>--------------<br>' + 'plate', data: {word_type: 'easy'} },
		{ stimulus: 'pig' + '<br>--------------<br>' + 'mud', data: {word_type: 'easy'} },	
		{ stimulus: 'skip' + '<br>--------------<br>' + 'insect', data: {word_type: 'hard'} }
		];
So the choice is actually made on the stimulus for "Enter A, B..." instead of the stimulus that I actually want to redisplay later in the program. (The above word pairs).
This is the code:

     var study_procedure = {
		timeline: [study,jol,choice],
		timeline_variables: test_stimuli,		
		on_finish: function(data){		
			condTargets = jsPsych.data.get().filter(key_press: 65);//collect rows of data for 'a' or 
                'b' keypresses
			new_stimuli = ("stimulus: " + condTargets.select('stimulus').values);
		}
		};
		
	var restudy_retest_procedure = {
		timeline: study,
		timeline_variables: new_stimuli,		
		};

OK - so 1) I would expect with this code to see "stimulus:  Enter A to study again", but all I'm getting in the console is "stimulus:                                           " (blank).
2) how do I get to the stimulus from the study procedure, which is actually what I want to display again? This is not a looping function, because I want to show ALL of the study word pairs BEFORE redisplaying the chosen word pairs to restudy or pretest. I need to finish all word pairs before going back and finding the flagged word pairs to display again. 

So currently, the program runs and displays all word pairs, then ends, because the new_stimuli array is blank.

Thanks,
Rose

**Comment by vijaymarupudi** [comment]
Having dynamic experimental designs like these usually means forgoing timeline variables and embracing loops and functions. Something like this.

```javascript
let second_timeline = []; // Array.concat this with the full timeline.
for (let i = 0; i < new_stimuli.length; i++) {
  second_timeline.push({
    type:  "html-keyboard-response",
    stimulus: function() {
      return new_stimuli[i];
    }
  })
}
```

Then the stimulus will be evaluated at the time of the trial, when new_stimuli will have been assigned.

**Reply by rjenki03** [reply]
Thank you! I will try this.

**Reply by jkhartshorne** [reply]
@rjenki03 did that work?
